File: go/problems/binary_watch.go
package problems

// https://leetcode.com/problems/binary-watch/ - Backtracking just the hours part
var hourConstants = [4]int{1, 2, 4, 8}

func backtrackHours(hours *[]int, num int, index int, result int) {
	if result > 11 {
		return
	}

	if num == 0 || index > 4 {
		*hours = append(*hours, result)
		return
	}
	for i := index; i < 4; i++ {
		backtrackHours(hours, num - 1, i + 1, result + hourConstants[i])
	}
}

func GetHours(num int) []int {
	var hours []int
	for i := 0; i <= num; i++ {
		backtrackHours(&hours, i, 0, 0)
	}
	return hours
}

File: rust/src/problems/binary_watch.rs
// https://leetcode.com/problems/binary-watch/ - Backtracking just the hours part
const HOUR_INDICES: [i32; 4] = [1, 2, 4, 8];

fn backtrack_hours(hours: &mut Vec<i32>, index: i32, result: i32, num: i32) {
    if result > 11 {
        return;
    }
    if num == 0 || index > 4 {
        hours.push(result);
        return;
    }
    for i in index..4 {
        backtrack_hours(hours, i + 1, result + HOUR_INDICES[i as usize], num - 1)
    }
}

pub fn get_hours(num: i32) -> Vec<i32> {
    let mut hours: Vec<i32> = Vec::new();
    for i in 0..=num {
        backtrack_hours(&mut hours, 0, 0, i);
    }
    return hours;
}
